src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java

```diff
@@ -0,0 +1,595 @@
+                && ItemStackUtils.isEmpty(((TileEntityFurnace) entity).getStackInSlot(ORE_SLOT))))
+        {
+            walkTo = null;
+            return START_WORKING;
+        }
+        walkTo = null;
+
+        gatherOreFromFurnaceWithChance((TileEntityFurnace) entity);
+
+        if (!ItemStackUtils.isEmpty(((TileEntityFurnace) entity).getStackInSlot(ORE_SLOT)))
+        {
+            if (!InventoryUtils.hasItemInItemHandler(new InvWrapper(worker.getInventoryCitizen()), TileEntityFurnace::isItemFuel))
+            {
+                walkTo = null;
+                return SMELTER_GET_FIREWOOD;
+            }
+
+            InventoryUtils.transferItemStackIntoNextFreeSlotInItemHandlers(
+                    new InvWrapper(worker.getInventoryCitizen()),
+                    InventoryUtils.findFirstSlotInItemHandlerWith(new InvWrapper(worker.getInventoryCitizen()), TileEntityFurnace::isItemFuel),
+                    new InvWrapper((TileEntityFurnace) entity));
+        }
+
+        incrementActionsDone();
+        setDelay(STANDARD_DELAY);
+        return START_WORKING;
+    }
+
+    private void gatherOreFromFurnaceWithChance(final TileEntityFurnace furnace)
+    {
+        final ItemStack ingots = new InvWrapper(furnace).extractItem(RESULT_SLOT, STACKSIZE, false);
+        final int multiplier = ((BuildingSmeltery) getOwnBuilding()).ingotMultiplier(worker.getCitizenData().getLevel(), worker.getRandom());
+        int amount = ingots.getCount() * multiplier;
+
+        while(amount > 0)
+        {
+            final ItemStack copyStack = ingots.copy();
+            if(amount  < ingots.getMaxStackSize())
+            {
+                copyStack.setCount(amount);
+            }
+            else
+            {
+                copyStack.setCount(ingots.getMaxStackSize());
+            }
+            amount -= copyStack.getCount();
+
+            final ItemStack resultStack = InventoryUtils.addItemStackToItemHandlerWithResult(new InvWrapper(worker.getInventoryCitizen()), ingots);
+            if(!ItemStackUtils.isEmpty(resultStack))
+            {
+                resultStack.setCount(resultStack.getCount() + amount / multiplier);
+                new InvWrapper(furnace).setStackInSlot(RESULT_SLOT, resultStack);
+                return;
+            }
+            worker.addExperience(BASE_XP_GAIN);
+        }
+    }
+
+    private AIState gatherOreFromBuilding()
+    {
+        if (needsCurrently == null)
+        {
+            needsCurrently = EntityAIWorkSmelter::isSmeltableOre;
+        }
+
+        final BlockPos pos = getOwnBuilding().getTileEntity().getPositionOfChestWithItemStack(needsCurrently);
+        if (pos == null)
+        {
+            return START_WORKING;
+        }
+
+        if (walkToBlock(pos))
+        {
+            return getState();
+        }
+
+        tryTransferFromPosToWorker(pos, needsCurrently);
+        setDelay(STANDARD_DELAY);
+        return START_WORKING;
+    }
+
+    private AIState smeltOre()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 7, down to the maximum allowed 6. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)